src/lshw.1
.\" auto-generated by docbook2man-spec $Revision: 1.2 $
.TH "LSHW" "1" "30 May 2010" "$Rev: 2179 $" ""
.SH NAME
lshw \- list hardware
.SH SYNOPSIS
.sp
\fBlshw\fR [ \fB-version\fR ] 
.sp
\fBlshw\fR [ \fB-help\fR ] 
.sp
\fBlshw\fR [ \fB-X\fR ] 
.sp
\fBlshw\fR [ \fB [ -html ]  [ -short ]  [ -xml ]  [ -json ]  [ -businfo ] \fR ]  [ \fB-dump \fIfilename\fB\fR ]  [ \fB-class \fIclass\fB\fR\fI...\fR ]  [ \fB-disable \fItest\fB\fR\fI...\fR ]  [ \fB-enable \fItest\fB\fR\fI...\fR ]  [ \fB-sanitize\fR ]  [ \fB-numeric\fR ]  [ \fB-quiet\fR ] 
.SH "DESCRIPTION"
.PP

\fBlshw\fR
is a small tool to extract detailed information on the hardware
configuration of the machine. It can report exact memory
configuration, firmware version, mainboard configuration, 
CPU version
and speed, cache configuration, bus speed, etc. on 
DMI-capable x86 or IA-64
systems and on some PowerPC
machines (PowerMac G4 is known to work).
.PP
It currently supports DMI (x86 and IA-64 only), OpenFirmware device tree (PowerPC only),
PCI/AGP, CPUID (x86), IDE/ATA/ATAPI, PCMCIA (only tested on x86), SCSI and USB\&.
.PP
.TP
\fB-version\fR
Displays the version of \fBlshw\fR and exits.
.TP
\fB-help\fR
Displays the available command line options and quits.
.TP
\fB-X\fR
Launch the X11 GUI (if available).
.TP
\fB-html\fR
Outputs the device tree as an HTML page.
.TP
\fB-xml\fR
Outputs the device tree as an XML tree.
.TP
\fB-json\fR
Outputs the device tree as a JSON object (JavaScript Object Notation).
.TP
\fB-short\fR
Outputs the device tree showing hardware paths, very much like the output of HP-UX\&'s \fBioscan\fR\&.
.TP
\fB-businfo\fR
Outputs the device list showing bus information, detailing SCSI, USB, IDE and PCI addresses.
.TP
\fB-dump \fIfilename\fB\fR
Dump collected information into a file (SQLite database).
.TP
\fB-class \fIclass\fB\fR
Only show the given class of hardware. \fIclass\fR can be found using \fBlshw -short\fR or \fBlshw -businfo\fR\&.
.TP
\fB-C \fIclass\fB\fR
Alias for \fB-class\fR \fIclass\fR\&.
.TP
\fB-enable \fItest\fB\fR
.TP
\fB-disable \fItest\fB\fR
Enables or disables a test. \fItest\fR can be \fBdmi\fR (for DMI/SMBIOS extensions), \fBdevice-tree\fR (for OpenFirmware device tree), \fBspd\fR (for memory Serial Presence Detect), \fBmemory\fR (for memory-size guessing heuristics), \fBcpuinfo\fR (for kernel-reported CPU detection), \fBcpuid\fR (for CPU detection), \fBpci\fR (for PCI/AGP access), \fBisapnp\fR (for ISA PnP extensions), \fBpcmcia\fR (for PCMCIA/PCCARD), \fBide\fR (for IDE/ATAPI), \fBusb\fR (for USB devices),\fBscsi\fR (for SCSI) or \fBnetwork\fR (for network interfaces detection).
.TP
\fB-quiet\fR
Don't display status.
.TP
\fB-sanitize\fR
Remove potentially sensitive information from output (IP addresses, serial numbers, etc.).
.TP
\fB-numeric\fR
Also display numeric IDs (for PCI and USB devices).
.SH "BUGS"
.PP
\fBlshw\fR currently does not detect 
Firewire(IEEE1394) devices.
.PP
Not all architectures supported by GNU/Linux are fully supported (e.g.
CPU detection).
.PP
"Virtual" SCSI interfaces used for SCSI emulation over IDE are not reported correctly yet.
.SH "NOTES"
.PP
\fBlshw\fR must be run as super user or it will only report
partial information.
.SH "FILES"
.PP
.TP
\fB/usr/local/share/pci.ids\fR
.TP
\fB/usr/share/pci.ids\fR
.TP
\fB/etc/pci.ids\fR
.TP
\fB/usr/share/hwdata/pci.ids\fR
A list of all known PCI ID's (vendors, devices, classes and subclasses).
.TP
\fB/proc/bus/pci/*\fR
Used to access the configuration of installed PCI busses and devices.
.TP
\fB/proc/ide/*\fR
Used to access the configuration of installed IDE busses and devices.
.TP
\fB/proc/scsi/*, /dev/sg*\fR
Used to access the configuration of installed SCSI devices.
.TP
\fB/dev/cpu/*/cpuid\fR
Used on x86 platforms to access CPU-specific configuration.
.TP
\fB/proc/device-tree/*\fR
Used on PowerPC platforms to access OpenFirmware configuration.
.TP
\fB/proc/bus/usb/*\fR
Used to access the configuration of installed USB busses and devices.
.TP
\fB/sys/*\fR
Used on 2.6 kernels to access hardware/driver configuration information.
.SH "EXAMPLES"
.PP
.TP
\fBlshw -short\fR
Lists hardware in a compact format.
.TP
\fBlshw -class disk -class storage\fR
Lists all disks and storage controllers in the system.
.TP
\fBlshw -html -class network\fR
Lists all network interfaces in HTML.
.TP
\fBlshw -disable dmi\fR
Don't use DMI to detect hardware.
.SH "SEE ALSO"
.PP
\fI/proc/*\fR, \fBlinuxinfo\fR(1), \fBlspci\fR(8), \fBlsusb\fR(8)
.SH "COPYING"
.PP
\fBlshw\fR is distributed under the GNU GENERAL PUBLIC LICENSE (GPL) version 2.
.SH "AUTHOR"
.PP
\fBlshw\fR is maintained by Lyonel Vincent
<lyonel@ezix.org>\&.
.SH "OTHER INFO"
.PP
The webpage for \fBlshw\fR is at 
 <URL:http://lshw.org/>


debian/changelog
lshw (02.16-2genesi0) maverick; urgency=low

  * Upstream lshw package built for the Efika MX53

 -- Chris Jenkins <chris.jenkins@genesi-usa.com>  Mon, 02 Jun 2014 21:17:35 -0500

lshw (02.16-2) unstable; urgency=low

  * Closes: #525217
    - lshw: segfauls on armel (! /proc/efi/systab)
  * Closes: #707152
    - lshw can miscount cpus on devicetree platforms

 -- Ghe Rivero <ghe@debian.org>  Thu, 27 Jun 2013 04:10:58 -0400

lshw (02.16-1) unstable; urgency=low

  * New upstream release

 -- Ghe Rivero <ghe@debian.org>  Sat, 30 Jun 2012 10:15:18 +0200

lshw (02.15-2) unstable; urgency=low

  * Relaxed depends on pciutils and usbutils. Closes: #540119

 -- Ghe Rivero <ghe.rivero@gmail.com>  Sun, 18 Dec 2011 13:26:10 +0100

lshw (02.15-1) unstable; urgency=low

  * [bbfc08f] Imported Upstream version 02.15
  * [927e643] Updated debian/watch file
  * [7ce089a] Clean some patches
  * [2ef1ea7] Added libsqlite3-dev as build-depend
  * [ea1f7f5] Bumped policy version
  * [2254715] New upstream release fixed FTBFS gcc4.5 (Closes: #565031)
  * [abac6bb] Added Spanish translation to .desktop.
    Thanks to Javier Fernández-Sanguino Peña (JFS) (Closes: #537904)
  * [ef94e6e] Added gksu launcher to .desktop.
    Thanks to Javier Fernández-Sanguino Peña (JFS) (Closes: #537903)
  * [ed92c99] Lintian clean again
  * [5e64767] Fixed manpage
  
 -- Ghe Rivero <ghe@debian.org>  Thu, 17 Feb 2011 10:10:08 +0100

lshw (02.14-1) unstable; urgency=low

  * New upstream release
  * Some lintian clean 
  * Standars Version 3.8.2 

 -- Ghe Rivero <ghe@debian.org>  Sat, 25 Jul 2009 15:30:52 +0200

lshw (02.13-2) unstable; urgency=low

  * Added patch sparc: FTBFS in sparc arch. (Closes: #480628)

 -- Ghe Rivero <ghe@debian.org>  Sun, 11 May 2008 16:09:55 +0200

lshw (02.13-1) unstable; urgency=low

  * New upstream release
    - Fixe lshw: no output for memory (Closes: #456896) 
  * Migrated dpatch to quilt
  * Clean lintian  (no more copyright-without-copyright-notice)

 -- Ghe Rivero <ghe@debian.org>  Sat, 10 May 2008 15:36:31 +0200

lshw (02.12.01-4) unstable; urgency=low

  * DudesConf2 release
  * Added patch 09-sysfs-hang
  * Added patch 10-privacy 

 -- Ghe Rivero <ghe@debian.org>  Sat, 03 May 2008 12:37:32 +0200

lshw (02.12.01-3) unstable; urgency=low

  * Now really updated Debian-Policy Standards-Version to 3.7.3
  * watch file working again. (Closes: #453535)
  * lintian cleans:
    - debian-rules-ignores-make-clean-error
    - debian-rules-sets-DH_COMPAT
    - dpatch-missing-description 

 -- Ghe Rivero <ghe@debian.org>  Sat,  2 Feb 2008 18:04:41 +0100

lshw (02.12.01-2) unstable; urgency=low

  * More fixes for gcc4.3. Closes: #462000
  * Updated Debian-Policy Standards-Version to 3.7.3

 -- Ghe Rivero <ghe@debian.org>  Wed, 23 Jan 2008 14:20:46 +0100

lshw (02.12.01-1) unstable; urgency=low

  * New upstream release
  * More fixes for gcc4.3 (Ubuntu Patch). Closes: #441569
  * Some Lintian cleans:
      Remove homepage from page description
  * Static man-page (no build-depends on docbook-utils)
  * Moved icon menu to /usr/share/pixmaps
  * Added /usr/share/pixmpas and /usr/share/applications to lshw-gtk.dir

 -- Ghe Rivero <ghe@debian.org>  Wed, 26 Dec 2007 13:22:07 +0100

lshw (02.11.01-2) unstable; urgency=low

  * s/gtk-lshw/lshw-gtk/g Closes: #443771
  * Some lintian cleans (new menu  Applications section)

 -- Ghe Rivero <ghe@debian.org>  Mon,  8 Oct 2007 17:29:45 +0200

lshw (02.11.01-1) unstable; urgency=low

  * New upstream release
  * Fixed ftbfs with g++-4.3/gcc-snapshot (07-gcc43.dpath). Closes: #441569
  * Explain why lshw -X fail in non X environment. Closes: #434644

 -- Ghe Rivero <ghe@debian.org>  Wed, 12 Sep 2007 09:54:48 +0200

lshw (02.10-4) unstable; urgency=low

  * Added Upstream Homepage to package description. Closes: #426698
  * Updated Debian-Policy Standards-Version to 3.7.2

 -- Ghe Rivero <ghe@debian.org>  Wed, 30 May 2007 13:02:26 +0200

lshw (02.10-3) unstable; urgency=low

  * New URL at debian/copyright. Closes: #405452 

 -- Ghe Rivero <ghe@debian.org>  Mon, 28 May 2007 18:09:22 +0200

lshw (02.10-2) unstable; urgency=low

  * New upstream release. Closes: #413800
  * New Maintainer mail address in control file

 -- Ghe Rivero <ghe@debian.org>  Fri, 25 May 2007 11:02:42 +0200

lshw (02.10-1) unstable; urgency=low

  * New upstream release. Closes: #405452
  * Remove lshw-common package. manuf.txt and oui.txt aren't used at all,
      and we now depend on pciutils and usbutils for pci.ids and usb.ids
      respectively (ubuntu patch)
  * New URL at debian/copyright. Closes: #405452
  * Fix FTBFS with GCC 4.3: missing #includes. Closes: #417381 
  
 -- Ghe Rivero <ghe@debian.org>  Thu, 03 May 2007 20:02:07 +0200

lshw (02.08.01-1) unstable; urgency=low

  * New upstream release. Closes: #360182
  * Added AMD64 to the known platforms that works properly. Closes: #349619

 -- Ghe Rivero <ghe@upsa.es>  Thu, 20 Jul 2006 00:35:07 +0200

lshw (02.06-3) unstable; urgency=low

  * "See you in GUA, Debconf-es2" Release
  * Change CXX flags on alpha, so no more segfaults.
  	Closes: #243752 and similar not reported to BTS.
  * Small fix on code for alpha (dpatch 06-alpha)
  * gxx transition
  * Dedicated to Endesa to switch off the electricity from my university
  	all the weekend :/
  * Updated DH_COMPAT=4
  * Removed ${shlibs:Depends} variable from lshw-common

 -- Ghe Rivero <ghe@upsa.es>  Wed, 21 Dec 2005 16:58:49 +0100

lshw (02.06-2) unstable; urgency=low

  * Updated rules files. Copy logy.xpm instead of move it.

 -- Ghe Rivero <ghe@upsa.es>  Mon, 17 Oct 2005 21:48:29 +0200

lshw (02.06-1) unstable; urgency=low

  * New upstream release

 -- Ghe Rivero <ghe@upsa.es>  Sun, 16 Oct 2005 23:59:50 +0200

lshw (02.05.01-3) unstable; urgency=low

  * Removed spurious files from source.
  * Make works with DEB_BUILD_OPTIONS.
  * Now can be compiled twice or more.

 -- Ghe Rivero <ghe@upsa.es>  Sun, 16 Oct 2005 23:11:27 +0200

lshw (02.05.01-2) unstable; urgency=low


src/lshw.cc
#include "hw.h"
#include "print.h"
#include "main.h"
#include "version.h"
#include "options.h"
#include "osutils.h"
#include "config.h"

#include <unistd.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <iostream>
#include <errno.h>

#ifndef NONLS
#include <locale.h>
#endif

__ID("@(#) $Id: lshw.cc 2433 2012-01-10 22:01:30Z lyonel $");

void usage(const char *progname)
{
  fprintf(stderr, "Hardware Lister (lshw) - %s\n", getpackageversion());
  fprintf(stderr, _("usage: %s [-format] [-options ...]\n"), progname);
  fprintf(stderr, _("       %s -version\n"), progname);
  fprintf(stderr, "\n");
  fprintf(stderr, _("\t-version        print program version (%s)\n"), getpackageversion());
  fprintf(stderr, _("\nformat can be\n"));
  fprintf(stderr, _("\t-html           output hardware tree as HTML\n"));
  fprintf(stderr, _("\t-xml            output hardware tree as XML\n"));
  fprintf(stderr, _("\t-short          output hardware paths\n"));
  fprintf(stderr, _("\t-businfo        output bus information\n"));
  if(getenv("DISPLAY") && exists(SBINDIR"/lshw-gtk"))
      fprintf(stderr, _("\t-X              use graphical interface\n"));
  fprintf(stderr, _("\noptions can be\n"));
#ifdef SQLITE
  fprintf(stderr, _("\t-dump OUTFILE   save hardware tree to a file\n"));
#endif
  fprintf(stderr, _("\t-class CLASS    only show a certain class of hardware\n"));
  fprintf(stderr, _("\t-C CLASS        same as '-class CLASS'\n"));
  fprintf(stderr, _("\t-c CLASS        same as '-class CLASS'\n"));
  fprintf(stderr,
    _("\t-disable TEST   disable a test (like pci, isapnp, cpuid, etc. )\n"));
  fprintf(stderr,
    _("\t-enable TEST    enable a test (like pci, isapnp, cpuid, etc. )\n"));
  fprintf(stderr, _("\t-quiet          don't display status\n"));
  fprintf(stderr, _("\t-sanitize       sanitize output (remove sensitive information like serial numbers, etc.)\n"));
  fprintf(stderr, _("\t-numeric        output numeric IDs (for PCI, USB, etc.)\n"));
  fprintf(stderr, "\n");
}


void status(const char *message)
{
  static size_t lastlen = 0;

  if(enabled("output:quiet") || disabled("output:verbose"))
    return;

  if (isatty(2))
  {
    fprintf(stderr, "\r");
    for (size_t i = 0; i < lastlen; i++)
      fprintf(stderr, " ");
    fprintf(stderr, "\r%s\r", message);
    fflush(stderr);
    lastlen = strlen(message);
  }
}


int main(int argc,
char **argv)
{

#ifndef NONLS
  setlocale (LC_ALL, "");
  bindtextdomain (PACKAGE, LOCALEDIR);
  bind_textdomain_codeset (PACKAGE, "UTF-8");
  textdomain (PACKAGE);
#endif

  disable("isapnp");

  disable("output:json");
  disable("output:db");
  disable("output:xml");
  disable("output:html");
  disable("output:hwpath");
  disable("output:businfo");
  disable("output:X");
  disable("output:quiet");
  disable("output:sanitize");
  disable("output:numeric");

// define some aliases for nodes classes
  alias("disc", "disk");
  alias("cpu", "processor");
  alias("lan", "network");
  alias("net", "network");
  alias("video", "display");
  alias("sound", "multimedia");
  alias("modem", "communication");

  if (!parse_options(argc, argv))
  {
    usage(argv[0]);
    exit(1);
  }

  while (argc >= 2)
  {
    bool validoption = false;

    if (strcmp(argv[1], "-version") == 0)
    {
      const char *latest = checkupdates();
      printf("%s\n", getpackageversion());
      if(latest)
      {
        if(strcmp(latest, getpackageversion()) != 0)
          fprintf(stderr, _("the latest version is %s\n"), latest);
      }
      exit(0);
    }

    if (strcmp(argv[1], "-help") == 0 || strcmp(argv[1], "--help") == 0)
    {
      usage(argv[0]);
      exit(0);
    }

    if (strcmp(argv[1], "-verbose") == 0)
    {
      disable("output:quiet");
      enable("output:verbose");
      validoption = true;
    }

    if (strcmp(argv[1], "-quiet") == 0)
    {
      disable("output:verbose");
      enable("output:quiet");
      validoption = true;
    }

    if (strcmp(argv[1], "-json") == 0)
    {
      enable("output:json");
      validoption = true;
    }

    if (strcmp(argv[1], "-xml") == 0)
    {
      enable("output:xml");
      validoption = true;
    }

    if (strcmp(argv[1], "-html") == 0)
    {
      enable("output:html");
      validoption = true;
    }

    if (strcmp(argv[1], "-short") == 0)
    {
      enable("output:hwpath");
      validoption = true;
    }

    if (strcmp(argv[1], "-businfo") == 0)
    {
      enable("output:businfo");
      validoption = true;
    }

    if (strcmp(argv[1], "-X") == 0)
    {
      enable("output:X");
      validoption = true;
    }

    if ((strcmp(argv[1], "-sanitize") == 0) ||
       (strcmp(argv[1], "-sanitise") == 0))
    {
      enable("output:sanitize");
      validoption = true;
    }

    if (strcmp(argv[1], "-numeric") == 0)
    {
      enable("output:numeric");
      validoption = true;
    }

    if(validoption)
    {	/* shift */
      memmove(argv+1, argv+2, (argc-1)*(sizeof(argv[0])));
      argc--;
    }
    else
    {
      usage(argv[0]);
      exit(1);
    }
  }

  if (argc >= 2)
  {
    usage(argv[0]);
    exit(1);
  }

/*  if(enabled("output:X")) execl(SBINDIR"/gtk-lshw", SBINDIR"/gtk-lshw", NULL); */

  if(enabled("output:X")) {
        execl(SBINDIR"/lshw-gtk", SBINDIR"/lshw-gtk", NULL);                                                                                                                                                    
        // Execl only returns if an error has occurred.
        if ( errno == ENOENT ) {
             fprintf(stderr, "ERROR: Sorry, cannot run the X11/GTK interface because %s/lshw-gtk\n is not available.\n", SBINDIR);                                                                               
             fprintf(stderr, "HINT: Install the lshw-gtk package in Debian.\n");                                                                                                                                 
        } else {
             fprintf(stderr, "ERROR: There was an error when trying to execute %s/lshw-gtk: %s\n", SBINDIR, strerror(errno));                                                                                    
             fprintf(stderr, "Will just print the information on screen\n");                                                                                                                                     
        }
  }

  if (geteuid() != 0)
  {
    fprintf(stderr, _("WARNING: you should run this program as super-user.\n"));
  }

  {
    hwNode computer("computer",
      hw::system);

    scan_system(computer);

    if (enabled("output:hwpath"))
      printhwpath(computer);
    else
    if (enabled("output:businfo"))
      printbusinfo(computer);
    else
    {
      if (enabled("output:json"))
        cout << computer.asJSON() << endl;
      else
      if (enabled("output:xml"))
        cout << computer.asXML();
      else
        print(computer, enabled("output:html"));
    }

    if(enabled("output:db"))
      computer.dump(getenv("OUTFILE"));
  }

  if (geteuid() != 0)
  {
    fprintf(stderr, _("WARNING: output may be incomplete or inaccurate, you should run this program as super-user.\n"));
  }


  return 0;
}
